src/api/app/services/trigger_controller_service/github_event_extractor.rb

```diff
@@ -0,0 +1,26 @@
+module TriggerControllerService
+  class GithubEventExtractor
+    ALLOWED_GITHUB_ACTIONS = ['opened', 'synchronize', 'closed'].freeze
+
+    def initialize(http_request)
+      @http_request = http_request
+      @payload = JSON.parse(@http_request.body.read)
+    end
+
+    def valid?
+      @http_request.env['HTTP_X_GITHUB_EVENT'] &&
+        @http_request.env['HTTP_X_GITHUB_EVENT'] == 'pull_request' &&
+        @payload['action'].in?(ALLOWED_GITHUB_ACTIONS)
+    end
+
+    # TODO: We also have to deal with wwwurlencoded parameters
+    # TODO: What happens when some of the keys are missing?
+    def add_payload_attributes_to_token(token)
```

Comment:
The payload attributes should be stored in the EventSubscriptions, not in the token. The token might have multiple EventSubscriptions. So if we pass the webhook payload to the `Token::Workflow` call method in the `create` action of the `TriggerController`, then we don't need to store that payload in the token. In the end, we'll do something like this to create the EventSubscriptions (see the payload there with `options[:something]`):

```ruby
    EventSubscription.create!(eventtype: 'Event::BuildSuccess',
                              receiver_role: 'watcher',
                              user: user,
                              channel: 'scm',
                              enabled: true,
                              token: self,
                              payload: {
                                scm: :github,
                                repository_owner: options[:repository_owner],
                                repository_name: options[:repository_name],
                                commit_sha: options[:commit_sha]
                              })
```